
#   Genomic Data File Formats

## Module Goals
- Become familiar with commonly used genomic data file formats
- Download necessary files for later analysis
- Practice UNIX commands to view and manipulate large genomic files

Genomic data can consist of sequences as well as information about sequence data. In order to store this information such that any program or pipeline can be applied to the data, there are defined genomic file types that follow specific formatting rules. While there are many types of genomic file types, here we will go over some of the most commonly used file formats which will be used in this project.


## Nucleotide Sequence Files

### FASTQ
- Stores the raw sequencing data which consists of nucleotide sequences and a quality score for each base in the sequence. Each nucleotide sequence, or read, represents a fragment of the genome which can be aligned to reference genome to identify variation or stitched together to generate an assembly. The quality score represents the probability of error at that base call (read more).

Example FASTQ line:
```text
@SRR1662283.1 1
TAAATAGTTCCCTTTGACAACACATGTGGCTGGTAGCGCCACTGTTTAGGCTCCACATCCAACGATCAAAGCCATCCATGTCCAGAGAGCCAAGC
+
ggggggggggggggggggggggggggggggeeeSdb^bcc]c``_ddOad^_bcZcbVbc^c`\\bPK^]Q[]YZ^b]]]`]cZ`Y\WXFa^_^`
```
- `@SRR1662283.1`: Each new read entry starts with a `@` followed by the sequence idenifier
- `TAAT...` : The raw nucleotide sequence
- `gggg...`: The per-base quality score

This example of a read comes from Illumina short-read sequencing, which is why it is only 100-bp. Long-reads, like the data used in this project, can range from 1 thousand to 1 million bp in length. 

### FASTA
- Stores nucleotide or amino acid sequences
- Starts each sequence with a `>` followed by the sequence name

Example FASTA file:
```text
>chr2L
CCCGTGCGCAGCCGGCAAGCACAGCATATTTGCGTGGCCGCTGTGAGTCTTTGTTATCTA
TAAGAATCATTTTAAGTTTCAGTTTGAATTTACATTTACCAGAATAAAGAGCGGCAGCTC
GTCAACTCCGACTAAGCGTCGTAAAACATATTTTTATTACTTACTCTGCCAGACCACAAG
GCTGACAGTTAAAAGAGGAAAGTTCCTCTCTACGACAGCATCGCAAAACGGTGCGATCAA
TTGTAGTGGCGACTGAAGCAGTGGGTGAGAAAAATCTCGATGAATAATCGATAGGGCGGT
CAGCTGTGACACACTAAAGTGAGGTCTTACTAGATGTAGCAAGAAGTAGCTAGAATATAC
CAGACTGGATGCAACCCTTATTAGCTGCACCCGATAAGTTTGGTGTAATTCGGTTATCGG
AGAAAAGTGTTGTTCAGCTGATCGTGGTTTCCGCCGATGGCCTACGCCTCGACTAATGTT
TCCGGTTTGCCTTCTTTTCGATCTGAACAGTCAACGCGTCGAGTTGCGAAAGTGCGTCGA
```

## Annotation Files
On their own, FASTA files are just strings of nucleotides. In order to locate and analyze regions of interest like genes or coding sequences within those genes, we need annotation files. Annotation files store information about the location of features in a genome, such as genes, enhancers, coding sequences, and repetitive elements.

### GTF/GFF
- Gene Transfer Format (GTF) files store information about gene structure (gene, exon, CDS, etc)
- General Feature Format (GFF) files store the same information as GTF, but generally include more detailed information about genomic features beyond those in genes.

GTF/GFF files have 9 tab-separated columns.

The 9 columns are:
1. Name of sequence (chromosome, scaffold, or contig)
2. Program or database that generated the feature	
3. The type of feature (gene, exon, CDS, repeat etc.)	
4. Starting position of the feature 
5. Ending position of the feature 
6. Score, sometimes represents a confidence score, depending on the program that generated it	
7. Strand of the feature, + is forward, - is reverse strand
8. Phase of the feature in a codon (1,2, or 0)	
9. Additional information about the feature, like the gene name

Columns 6,7, and 8 are '.' if unknown or not applicable

Example GFF file:
```text
2L      BDTNP1_TFBS_dl  TF_binding_site 5203    6186    .       .       .       ID=FBsf0000295671;Name=TFBS_dl_000001;Dbxref=FlyBase:FBsf0000295671
2L      FlyBase exon    21823   22687   .       -       .       Parent=FBtr0113008
2L      FlyBase gene    138384  140992  .       -       .       ID=FBgn0051976;Name=ovm;fullname=oven mitt;Alias=FBgn0031222,CG31976,CG11623;
2L      FlyBase CDS     11410   11518   .       -       2       Parent=FBtr0078170
```

### BED
- A much simpler format for storing genomic intervals.
- Can represent coordinates for anything: gene features, variants, or just a region of interest.
- Only requires 3 columns: a sequence name, a starting position, and an ending position.

Example of a simple bed file:
```text
2L  3900 6609
2L  7010    9047
2L  11593   12692    
```
---

## TO DO: Downloading and Working with the <i> Drosophila melanogaster </i> Reference Genome
A reference genome is a representative, well-annotated version of a species genome that serves as a standard for research and comparison. It is typically a consensus sequence assembled from the genetic material of multiple individuals.

The <i>Drosophila melanogaster</i> reference genome is one of the most complete and well-annotated metazoan genome assemblies, meaning we have detailed information about genomic sequences and gene locations.

An essential resource for accessing data from the extensive field of <i>Drosophila</i> research is called **FlyBase**. While youâ€™ll learn later how to search for gene and mutation information on FlyBase, for now, we'll use it to download the <i>D. melanogaster </i>  ISO1 reference genome data (ISO1 is the name of the reference strain).
### Step 1. Download FASTA
- On your browser, go to the [FlyBase homepage](https://flybase.org/).
- Find the **Downloads** button at the top of the page.
- In the dropdown menu, click **Genomes (FTP)**.
- Navigate by clicking this sequence of folders: Drosophila_melanogaster -> current -> fasta 
- Click the file starting with **dmel-all-chromosome-r6...** to download it.


### Step 2. View FASTA File
In your terminal, navigate to the directory where you have downloaded the reference FASTA file. Depending on the time of download, the release number may be different. For these instruction modules, we will use release 6 version 63 (r6.63).
```bash
# unzip the fasta
gunzip dmel-all-chromosome-r6.63.fasta.gz
# view the fasta
less dmel-all-chromosome-r6.63.fasta
```
*Note that after we unzip the file, it no longer has the .gz at the end of the name.

We can also use the command `head` to print out just the first 10 lines of a file:
```bash
head dmel-all-chromosome-r6.63.fasta
```
Output:
```text
>2L type=golden_path; loc=2L:1..23513712; ID=2L; dbxref=GB:AE014134,GB:AE014134,REFSEQ:NT_033779; MD5=b6a98b7c676bdaa11ec9521ed15aff2b; length=23513712; release=r6.63; species=Dmel;
CGACAATGCACGACAGAGGAAGCAGAACAGATATTTAGATTGCCTCTCATTTTCTCTCCCATATTATAGGGAGAAATATG
ATCGCGTATGCGAGAGTAGTGCCAACATATTGTGCTCTTTGATTTTTTGGCAACCCAAAATGGTGGCGGATGAACGAGAT
GATAATATATTCAAGTTGCCGCTAATCAGAAATAAATTCATTGCAACGTTAAATACAGCACAATATATGATCGCGTATGC
GAGAGTAGTGCCAACATATTGTGCTAATGAGTGCCTCTCGTTCTCTGTCTTATATTACCGCAAACCCAAAAAGACAATAC
ACGACAGAGAGAGAGAGCAGCGGAGATATTTAGATTGCCTATTAAATATGATCGCGTATGCGAGAGTAGTGCCAACATAT
```

---

## Annotation Files

### Step 3. Download GTF File
- Going back to FlyBase, click **<<-Parent Directory** to navigate back to the **current** folder.
- Click the folder **gtf**
- Click the file starting with **dmel-all-r6...** to download it.
- Unzip and use `head` to view the first few lines of the GTF. 
```bash
# unzip file
gunzip dmel-all-r6.63.gtf.gz
# view file
head dmel-all-r6.63.gtf
```
Output:
```text
X	FlyBase	gene	19961297	19969323	.	+	.	gene_id "FBgn0031081"; gene_symbol "Nep3";
X	FlyBase	mRNA	19961689	19968479	.	+	.	gene_id "FBgn0031081"; gene_symbol "Nep3"; transcript_id "FBtr0070000"; transcript_symbol "Nep3-RA";
X	FlyBase	5UTR	19961689	19961845	.	+	.	gene_id "FBgn0031081"; gene_symbol "Nep3"; transcript_id "FBtr0070000"; transcript_symbol "Nep3-RA";
X	FlyBase	exon	19961689	19961845	.	+	.	gene_id "FBgn0031081"; gene_symbol "Nep3"; transcript_id "FBtr0070000"; transcript_symbol "Nep3-RA";
X	FlyBase	exon	19963955	19964071	.	+	.	gene_id "FBgn0031081"; gene_symbol "Nep3"; transcript_id "FBtr0070000"; transcript_symbol "Nep3-RA";
X	FlyBase	exon	19964782	19964944	.	+	.	gene_id "FBgn0031081"; gene_symbol "Nep3"; transcript_id "FBtr0070000"; transcript_symbol "Nep3-RA";
X	FlyBase	exon	19965006	19965126	.	+	.	gene_id "FBgn0031081"; gene_symbol "Nep3"; transcript_id "FBtr0070000"; transcript_symbol "Nep3-RA";
X	FlyBase	exon	19965197	19965511	.	+	.	gene_id "FBgn0031081"; gene_symbol "Nep3"; transcript_id "FBtr0070000"; transcript_symbol "Nep3-RA";
X	FlyBase	exon	19965577	19966071	.	+	.	gene_id "FBgn0031081"; gene_symbol "Nep3"; transcript_id "FBtr0070000"; transcript_symbol "Nep3-RA";
X	FlyBase	exon	19966183	19967012	.	+	.	gene_id "FBgn0031081"; gene_symbol "Nep3"; transcript_id "FBtr0070000"; transcript_symbol "Nep3-RA";
```

### Step 4. More UNIX commands to work with genomic data
Your new FASTA and GTF files contain a lot of information, but we can use built-in UNIX tools to search for patterns and pull out information we may need. Here we will go over some basic ways to use `grep` and `awk` commands to do so. 

## Using `grep`
The `grep` command can search for a specified text pattern in a file. 

Basic structure:

**grep "search_pattern" input-file**

Lets say we want to view lines in our GTF file related to the gene <i> speck </i>:
```bash
grep "speck" dmel-all-r6.63.gtf
```
Output:
```text
2R	FlyBase	gene	24139563	24157762	.	+	.	gene_id "FBgn0287831"; gene_symbol "speck";
2R	FlyBase	mRNA	24139563	24157762	.	+	.	gene_id "FBgn0287831"; gene_symbol "speck"; transcript_id "FBtr0072254"; transcript_symbol "speck-RA";
2R	FlyBase	5UTR	24139563	24139861	.	+	.	gene_id "FBgn0287831"; gene_symbol "speck"; transcript_id "FBtr0072254"; transcript_symbol "speck-RA";
2R	FlyBase	5UTR	24155804	24155842	.	+	.	gene_id "FBgn0287831"; gene_symbol "speck"; transcript_id "FBtr0072254"; transcript_symbol "speck-RA";
2R	FlyBase	exon	24139563	24139861	.	+	.	gene_id "FBgn0287831"; gene_symbol "speck"; transcript_id "FBtr0072254"; transcript_symbol "speck-RA";
2R	FlyBase	exon	24155804	24155977	.	+	.	gene_id "FBgn0287831"; gene_symbol "speck"; transcript_id "FBtr0072254"; transcript_symbol "speck-RA";
2R	FlyBase	exon	24156897	24157049	.	+	.	gene_id "FBgn0287831"; gene_symbol "speck"; transcript_id "FBtr0072254"; transcript_symbol "speck-RA";
2R	FlyBase	exon	24157114	24157762	.	+	.	gene_id "FBgn0287831"; gene_symbol "speck"; transcript_id "FBtr0072254"; transcript_symbol "speck-RA";
2R	FlyBase	start_codon	24155843	24155845	.	+	0	gene_id "FBgn0287831"; gene_symbol "speck"; transcript_id "FBtr0072254"; transcript_symbol "speck-RA";
```


## Using `awk`
The `awk` command can also be useful for identifying patterns and manupulating lines in a file. 

Basic structure:

**awk 'pattern {action}' input-file**

`awk` splits each line by tabs or spaces and assigns each column to a variable:

 Variables
- $1 = first column
- $2 = second column
- $5 = fifth column

 Which we can compare using operators:
- == equals
- != not equal
- \> greater than
- < less than

Examples:
```bash
# print out lines in our GFF file where the 3rd column ($3) is 'exon'
awk '$3 == "exon" {print}' dmel-all-r6.63.gtf
# if the 3rd column is 'exon', print out only the 5th column
awk '$3 == "exon" {print $5}' dmel-all-r6.63.gtf
```

---

## Piping Commands
You can use the pipe character `|` to chain commands together:
```bash
# get only gene annotations from GTF and print out first 10 genes
awk '$3 == "gene" {print}' dmel-all-r6.63.gtf | head
```
Output:
```text
X	FlyBase	gene	19961297	19969323	.	+	.	gene_id "FBgn0031081"; gene_symbol "Nep3";
X	FlyBase	gene	20025099	20025170	.	+	.	gene_id "FBgn0052826"; gene_symbol "tRNA:Pro-CGG-1-1";
X	FlyBase	gene	20051294	20052519	.	+	.	gene_id "FBgn0031085"; gene_symbol "CG9570";
X	FlyBase	gene	20094398	20095767	.	+	.	gene_id "FBgn0062565"; gene_symbol "Or19b";
X	FlyBase	gene	20133579	20138878	.	+	.	gene_id "FBgn0031088"; gene_symbol "CG15322";
X	FlyBase	gene	20141819	20143188	.	-	.	gene_id "FBgn0041626"; gene_symbol "Or19a";
X	FlyBase	gene	20110110	20110575	.	-	.	gene_id "FBgn0040784"; gene_symbol "karr";
X	FlyBase	gene	20091428	20092408	.	-	.	gene_id "FBgn0031086"; gene_symbol "fd19B";
X	FlyBase	gene	20000968	20002096	.	-	.	gene_id "FBgn0031082"; gene_symbol "lky";
X	FlyBase	gene	19958393	19959403	.	-	.	gene_id "FBgn0031080"; gene_symbol "CG12655";
```

What if we want to see all the sequence headers in our FASTA file? We know that sequence headers always start with a `>` symbol:
```bash
grep ">" dmel-all-chromosome-r6.63.fasta | head
```
Output:
```text
>2L type=golden_path; loc=2L:1..23513712; ID=2L; dbxref=GB:AE014134,GB:AE014134,REFSEQ:NT_033779; MD5=b6a98b7c676bdaa11ec9521ed15aff2b; length=23513712; release=r6.63; species=Dmel;
>2R type=golden_path; loc=2R:1..25286936; ID=2R; dbxref=GB:AE013599,GB:AE013599,REFSEQ:NT_033778; MD5=2ecce4ca1c0106bb7c63a78b93ab49ba; length=25286936; release=r6.63; species=Dmel;
>3L type=golden_path; loc=3L:1..28110227; ID=3L; dbxref=GB:AE014296,GB:AE014296,REFSEQ:NT_037436; MD5=3c3ea06b22af8cc59809dbf8d154791e; length=28110227; release=r6.63; species=Dmel;
>3R type=golden_path; loc=3R:1..32079331; ID=3R; dbxref=GB:AE014297,GB:AE014297,REFSEQ:NT_033777; MD5=420540d26d86fbf14185d2f2d68af9c4; length=32079331; release=r6.63; species=Dmel;
>4 type=golden_path; loc=4:1..1348131; ID=4; dbxref=GB:AE014135,GB:AE014135,REFSEQ:NC_004353; MD5=8dbdf26f38049dd9df4076b6d480ae61; length=1348131; release=r6.63; species=Dmel;
>X type=golden_path; loc=X:1..23542271; ID=X; dbxref=GB:AE014298,GB:AE014298,REFSEQ:NC_004354; MD5=f7228bbf9f8fa0af457bc2fb7e396e4d; length=23542271; release=r6.63; species=Dmel;
>Y type=golden_path; loc=Y:1..3667352; ID=Y; dbxref=GB:CP007106,GB:CP007106,REFSEQ:NC_024512; MD5=0938dd2253a423799e7938dc892f1c04; length=3667352; release=r6.63; species=Dmel;
>2Cen_mapped_Scaffold_10_D1684 type=golden_path; loc=2Cen_mapped_Scaffold_10_D1684:1..19956; ID=2Cen_mapped_Scaffold_10_D1684; dbxref=GB:CP007071,GB:CP007071,REFSEQ:NW_007931073; MD5=b71b59981d0bc7d367792f99638e8ff5; length=19956; release=r6.63; species=Dmel;
>2Cen_mapped_Scaffold_43_D1668 type=golden_path; loc=2Cen_mapped_Scaffold_43_D1668:1..44411; ID=2Cen_mapped_Scaffold_43_D1668; dbxref=GB:CP007072,GB:CP007072,REFSEQ:NW_007931074; MD5=8f5e867a2b263eb401acf5a10a1a8f19; length=44411; release=r6.63; species=Dmel;
>2R2_mapped_Scaffold_56_D1828 type=golden_path; loc=2R2_mapped_Scaffold_56_D1828:1..13157; ID=2R2_mapped_Scaffold_56_D1828; dbxref=GB:CP007073,GB:CP007073,REFSEQ:NW_007931075; MD5=5aedd39059ed537a31be2767e816855a; length=13157; release=r6.63; species=Dmel;
```

**MINI FLY GENETICS LESSON!** 
The *Drosophila melanogaster* genome consists of four pairs of chromosomes: the sex chromosomes (X and Y) and three pairs of autosomes (chromosomes 2, 3, and 4). Chromosomes 2 and 3 have large, heterochromatic centromeres located roughly in the middle, which divide each into a left and right armâ€”referred to as 2L/2R and 3L/3R, respectively.

As you may have noticed from the first seven lines of output in the previous `grep` command, the reference genome includes a separate sequence entry for each major chromosome arm (2L, 2R, 3L, 3R), along with chromosomes 4, X, and Y. These are considered the main chromosomes, and we typically focus on them during genomic analyses. Any sequences listed after these are usually unplaced or minor contigs.




## Directing output of commands into a file
So far, the output of our commands has gone to the terminal, or standard output. If we want to save the output into a file we can direct it using the `>` character.

**command input_file > output_file**

Example:
```bash
# create a file called 'genes.txt' containing gene features from GTF
awk '$3 == "gene" {print}' dmel-all-r6.63.gtf > genes.txt
```
If you check your new file genes.txt with `less`, you should now have a file containing the coordinates of all genes annotated in the <i> D. melanogaster </i> reference genome.


## Practice Suggestions
1. Use `awk`, `grep`, and piping to print out only the exon coordinates of a specific gene (example genes: *Polr1B*, *deltaCOP*, *Cyp4d14*) 
2. Use `awk`, `grep`, and piping to print out only the gene coordinates on chromosome 3R.
