# Genome Assembly

Once your sequencing reads have been filtered for quality and length, the next step is to **assemble them into a draft genome**. Genome assembly is the process of stitching together overlapping sequencing reads to reconstruct the original genome. For Oxford Nanopore reads, which are long but can be error-prone, assemblers designed specifically for long reads (such as [Hifiasm](https://www.nature.com/articles/s41592-020-01056-5) or [Flye](https://www.nature.com/articles/s41587-019-0072-8)) are ideal.

In this example, we use **Hifiasm**, a fast and accurate assembler optimized for PacBioHiFi and, more recently, ONT long-read data.

## SLURM Job Script to Run Hifiasm
Some tasks, such as genome assembly, requires more resources and time, so if you are operating on an high-performance computing cluster (HPC) that uses SLURM to manage jobs, create a script using nano (ex. `nano hifiasm.sh`) and paste the following job specifications (#SBATCH) at the top of the script. This SLURM job will allocate resources and manages the assembly job on an HPC. 

```bash
#!/bin/bash

#SBATCH --job-name=hifiasm_BL2969
#SBATCH --time=6:00:00
#SBATCH --mem=100G
#SBATCH --ntasks-per-node=48
#SBATCH --error=job.%J.hifiasm.err
#SBATCH --out=job.%J.hifiasm.out
```

This sets up a high-performance computing (HPC) job that:

*   Allocates 100 GB of memory and 48 CPU threads
*   Runs for a maximum of 6 hours
*   Logs error and output messages

Then paste the command to load HiFiasm modules or activate the environment with conda:

## Run the Assembly
```bash
module load GCCcore/13.2.0 hifiasm/0.24.0

hifiasm --ont -o BL2969_hifiasm -t 48 ../BL2969_filt.fastq
```
- `--ont`: specifies that the reads are Oxford Nanopore
- `-o`: output prefix
- `-t`: number of threads
- Input: the filtered FASTQ file

## Output
The output will include:
- Contigs in FASTA format (`.fa`)
- Assembly graph files
- Log files

These **contigs** represent your preliminary draft genome â€” continuous sequences assembled from overlapping reads. However, this assembly may still contain **contamination**, **assembly errors**, or **redundant sequences**.


