# Optional: Decontamination 

The following Decontamination and Completeness Assessment sections are optional steps that are typically done when creating a genome assembly. However, tools like BLAST and BUSCO can require a significant amount of space and computing time since they use large databases to compare the genome against. On a high-powered computer or a computing cluster this may be more feasible, however, for solo work on a personal computer it may take a lot of resources and time. If these steps are not performed, you can still continue on to the Variant Detection modules with the draft genome assembly.

## Decontamination of Draft Assembly

Although the sample used for DNA extraction came from 150 *Drosophila melanogaster* females, contamination from microorganisms (especially gut bacteria) is quite common. These contaminants can end up being assembled into the genome.

To identify and remove contaminant contigs, we use **BlobTools**, which integrates:
- A BLAST search of contigs against the NCBI nucleotide database (for taxonomic classification)
- Read coverage (via read mapping back to contigs)
- GC content

This enables us to identify contigs with unexpected taxonomy, unusual GC content, or abnormal coverage.

### BLASTN of Contigs


export BLASTDB=/scratch/data/bio/blast/

blastn -query BL1282_hifiasm_contigs.fa -db nt \
  -outfmt '6 qseqid staxids bitscore std' \
  -num_threads 48 -max_target_seqs 1 -max_hsps 1 \
  -evalue 1e-25 -out BL1282.blast.out
```

### Read Mapping
Blobtools requires that the reads used to generate the genome assembly be mapped back to the assembly itself to identify contigs that have higher coverage than others.

```bash
minimap2 -ax map-ont BL662_clean_contigs.fna BL662_filt.fastq \
  | samtools sort -@48 -O BAM -o BL662_reads2contigs.bam -
samtools index BL662_reads2contigs.bam
```

### BlobTools Pipeline
Blobtools then takes the assembly, BLAST output, and read alignment BAM to create a database, which we can produce visual representations of the data from as 'blob plots'

```bash
module load GCC/7.3.0-2.30 OpenMPI/3.1.1 BlobTools/20180528-Python-2.7.15

blobtools create -i ZI457N_flye.fa -b ZI457N_reads2asm.bam \
  -t ZI457N.blast.out -o ZI457N

blobtools view -i ZI457N.blobDB.json -o ZI457N_blob
blobtools plot -i ZI457N.blobDB.json
```

### Filter Out Non-*Drosophila* Contigs
You will probably see that some contigs correspond to the smaller genomes of microbes like yeast or parasites like Wolbachia. We want to filter those out to keep only those that are classified as arthropods (what flies are) or no-hit as these could represent novel sequences that BLAST couldn't identify because they are not in the database.
```bash
awk '$6 == "Arthropoda" || $6 == "no-hit"' ZI457N_blob.ZI457N.blobDB.table.txt | \
  cut -f1 > ZI457N_contigs2keep.txt
blobtools seqfilter -i ZI457N_flye.fa -l ZI457N_contigs2keep.txt
```

This retains only contigs classified as *Arthropoda* or not classified at all (`no-hit`).

