# Evaluating Genome Assembly Quality

Once we have a cleaned genome assembly, we assess its quality, typically by three characteristics:

## 1. Contiguity
- Measured by statistics like **N50**, which is the length such that 50% of the assembly is contained in contigs of that length or longer.
- Tools like **QUAST** calculate N50, L50, total length, number of contigs, and more.
```bash
quast -o strain_contigs_quast assembly.fa

```

## 2. Completeness
- How much of the expected gene content is present?
- This is measured using **BUSCO**, which looks for a set of conserved genes expected in the target lineage (ex. Diptera).

```bash
busco -i scaffs.fa -m genome -l lineage database -c 24 -o dip_busco --offline
```

## 3. Accuracy
- While not directly addressed in this pipeline, accuracy can be estimated with polishing tools (e.g., Medaka, Racon) or by comparing to a reference genome.

---