## Variant Discovery

Within a species, there can be a wide range of genetic variation, from single nucleotide polymorphisms (SNPs) to large-scale genomic rearrangements known as structural variants (SVs). SVs include duplications, deletions, insertions, and inversions of sequences larger than 100 base pairs.

Thanks to the long history of *Drosophila* genetics research, many of the visible phenotypes observed in the 11 *Drosophila* strains examined in this project have already been linked to specific genes â€” and in some cases, to particular SNPs or SVs. However, the sequencing data generated here provides a unique opportunity to apply modern genomics tools to further investigate the molecular nature of these mutations. For phenotypes with uncharacterized genetic causes, this also opens the door to discovering novel candidate mutations.

These modules will guide you through two variant-calling approaches (read mapping and whole-genome alignment) to identify genetic variants and pinpoint mutations in genes associated with visible phenotypes.

- 3.1 - Read Mapping

- 3.2 - Whole Genome Alignment

- 3.3 - Repeat Annotation

- 3.4 - Candidate Mutation Discovery